lib/firmware/nano.c

```diff
@@ -0,0 +1,86 @@
+#include <string.h>
+#include <stdio.h>
+#include <inttypes.h>
+#include "keepkey/firmware/nano.h"
+#include "trezor/crypto/blake2b.h"
+
+static uint8_t const NANO_BLOCK_HASH_PREAMBLE[32] = {
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+    0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x06,
+};
+
+bool nano_path_mismatched(const CoinType *coin,
+                          const uint32_t *address_n,
+                          const uint32_t address_n_count)
+{
+    bool mismatch = false;
+
+    // m : no path
+    if (address_n_count == 0) {
+        return false;
+    }
+
+    // m/44' : BIP44
+    // m / purpose' / bip44_account_path' / account'
+    if (address_n[0] == (0x80000000 + 44)) {
+        mismatch |= address_n_count < 3;
+        mismatch |= address_n_count > 1 && (address_n[1] != coin->bip44_account_path);
+        mismatch |= address_n_count > 2 && (address_n[2] & 0x80000000) == 0;
+        mismatch |= address_n_count > 3 && (address_n[3] & 0x80000000) == 0;
```

Comment:
What are `address_n[3]` and `address_n[4]` for in Nano derivation paths?

Reply:
Nothing specific. The Ledger wallet is completely un-opinionated about the path components and it's up to the wallet (or user) to choose the paths for the accounts. So the idea here was to be somewhat interoperable with the Ledger implementation, should the users want to use the extra path components.

Reply:
Our attitude toward it is to strongly encourage one particular layout, and allow, but warn against using others. This encourages using standardized layouts since those are the ones with the nice human-readable descriptions, and gives others the raw bip32 path & warn on receive. It also allows people to migrate from nonstandard layouts if they need to, and helps users do the Right Thing (TM) that gives them compatibility with most other wallet implementations. Being too flexible here IMO leads to misplacement / apparent loss of funds, and difficult wallet UX.

Reply:
Yeah, agreed.. it probably is the wrong place to introduce the flexibility.. I'll give this a do-over in the evening.

Reply:
Fixed in 958081b

Reply:
ty!